<a href="https://colab.research.google.com/github/neha-palak/cv-project/blob/main/CV_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install torch torchvision timm opencv-python matplotlib

# Set-up

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import os

In [ ]:
# os.rename('/content/drive/MyDrive/Khana Dataset',
#           '/content/drive/MyDrive/Khana_Dataset')

base_path = '/content/drive/MyDrive/Khana_Dataset'

os.listdir(base_path)

['taxonomy.csv', 'labels.txt', 'dataset.tar.gz']

In [ ]:
import tarfile

tar = tarfile.open(f'{base_path}/dataset.tar.gz')
tar.extractall('/content/')
tar.close()

/tmp/ipykernel_10919/3388360630.py:4: DeprecationWarning: Python 3.14 will, by default, filter extracted tar archives and reject files or modify their metadata. Use the filter argument to control this behavior.
  tar.extractall('/content/')


In [ ]:
print(os.listdir('/content'))
amount = len(os.listdir('/content/khana'))
print(amount)
print(os.listdir('/content/khana'))

['.config', 'khana', 'drive', 'sample_data']
82
['garlic naan', 'dal khichdi', 'gujhia', 'pani puri', 'chicken wings', 'kulfi', 'phirni', 'chole bhature', 'pongal', 'falooda', 'solkadhi', 'chicken pizza', 'pepperoni pizza', 'paneer pizza', 'kaju katli', 'sabudana khichdi', 'seekh kebab', 'dhokla', 'khakhra', 'balushahi', 'masala papad', 'samosa', 'ghevar', 'paniyaram', 'navratan korma', 'garlic bread', 'puri bhaji', 'dabeli', 'onion pakoda', 'murukku', 'papdi chaat', 'idiyappam', 'rajma chawal', 'boondi laddu', 'aloo methi', 'kheer', 'chapati', 'gajar ka halwa', 'chaas', 'hara bhara kabab', 'medu vada', 'neer dosa', 'biryani', 'chana masala', 'chivda', 'idli', 'modak', 'mysore pak', 'steamed momo', 'patrode', 'margherita pizza', 'besan laddu', 'poha', 'uttapam', '._.DS_Store', 'set dosa', 'moong dal halwa', 'thukpa', 'aloo gobi', 'fish curry', '.DS_Store', 'sev puri', 'thali', 'anda curry', 'gulab jamun', 'masala dosa', 'aloo paratha', 'sabudana vada', 'aloo mutter', 'banana chips', 'a

In [ ]:
for item in os.listdir('/content/khana'):
    if item.startswith('.'):
        os.system(f'rm -rf "/content/khana/{item}"')

In [ ]:
fixed_amount = len(os.listdir('/content/khana'))
print(fixed_amount)

80


In [ ]:
from PIL import ImageFile
ImageFile.LOAD_TRUNCATED_IMAGES = True

In [ ]:
# adding jpg extension
for cls in os.listdir('/content/khana'):
    cls_path = os.path.join('/content/khana', cls)

    if not os.path.isdir(cls_path):
        continue

    for file in os.listdir(cls_path):
        file_path = os.path.join(cls_path, file)

        # skip if already has extension
        if '.' not in file:
            new_path = file_path + '.jpg'
            os.rename(file_path, new_path)

In [ ]:
print(os.listdir('/content/khana/biryani')[:5])

['c7b8306554987118c6ed7eb02f6c6aad.jpg', 'amcgb3uvnqsmrp5loe1a.jpg', '9e148aa15084a144898de1ab1d485b2a.jpg', 'hoan9uexmioppvebv4fs.jpg', 'nlq3gqbnvvuuv9rwg2hb.jpg']


In [ ]:
import shutil

shutil.copytree('/content/khana', '/content/drive/MyDrive/your_folder_name')

FileExistsError: [Errno 17] File exists: '/content/drive/MyDrive/khana'

pipeline

In [ ]:
# import os, torch, torch.nn as nn
# from torchvision import datasets, transforms, models
# from torch.utils.data import DataLoader, WeightedRandomSampler
# from torch.optim import AdamW
# from torch.optim.lr_scheduler import OneCycleLR
# import numpy as np

# DATASET_PATH = '/content/khana'
# IMG_SIZE = 224
# BATCH_SIZE = 32
# NUM_CLASSES = 80
# DEVICE = 'cuda' if torch.cuda.is_available() else 'cpu'
# print(f"Using: {DEVICE}")

In [ ]:
# train_tfm = transforms.Compose([
#     transforms.RandomResizedCrop(IMG_SIZE, scale=(0.6, 1.0)),
#     transforms.RandomHorizontalFlip(),
#     transforms.ColorJitter(brightness=0.3, contrast=0.3, saturation=0.3, hue=0.1),
#     transforms.RandomRotation(20),
#     transforms.ToTensor(),
#     transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225]),
# ])
# val_tfm = transforms.Compose([
#     transforms.Resize(256),
#     transforms.CenterCrop(IMG_SIZE),
#     transforms.ToTensor(),
#     transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225]),
# ])

# full_ds = datasets.ImageFolder(DATASET_PATH)
# n = len(full_ds)
# n_val = int(0.2 * n)
# n_train = n - n_val
# train_ds, val_ds = torch.utils.data.random_split(
#     full_ds, [n_train, n_val],
#     generator=torch.Generator().manual_seed(42)
# )
# train_ds.dataset.transform = train_tfm
# val_ds.dataset.transform   = val_tfm

# # Class-balanced sampler (helps with imbalanced classes)
# targets = [full_ds.targets[i] for i in train_ds.indices]
# class_counts = np.bincount(targets)
# weights = 1.0 / class_counts[targets]
# sampler = WeightedRandomSampler(weights, len(weights))

# # Change num_workers from 2 to 4, and add pin_memory + prefetch
# train_loader = DataLoader(train_ds, batch_size=BATCH_SIZE, sampler=sampler,
#                           num_workers=2, pin_memory=True)
# val_loader   = DataLoader(val_ds, batch_size=BATCH_SIZE, shuffle=False,
#                           num_workers=2, pin_memory=True)
# print(f"Train: {n_train}  Val: {n_val}  Classes: {len(full_ds.classes)}")

In [ ]:
# model = models.efficientnet_b4(weights='IMAGENET1K_V1')
# model.classifier[1] = nn.Linear(model.classifier[1].in_features, NUM_CLASSES)
# model = model.to(DEVICE)

# # Alternatively: ConvNeXt-Small (usually slightly better)
# # model = models.convnext_small(weights='IMAGENET1K_V1')
# # model.classifier[2] = nn.Linear(768, NUM_CLASSES)

In [ ]:
# !nvidia-smi


In [ ]:
# import warnings
# warnings.filterwarnings('ignore')
# import torch.backends.cudnn as cudnn
# cudnn.benchmark = True  # speeds up conv ops for fixed input sizes
# criterion = nn.CrossEntropyLoss(label_smoothing=0.1)
# optimizer = AdamW(model.parameters(), lr=1e-4, weight_decay=1e-2)

# EPOCHS = 10
# scheduler = OneCycleLR(optimizer, max_lr=3e-4,
#                        steps_per_epoch=len(train_loader), epochs=EPOCHS)

# ACCUMULATION_STEPS = 4  # effective batch = 32 * 4 = 128

# def train_epoch():
#     model.train(); total, correct = 0, 0
#     optimizer.zero_grad()
#     for i, (imgs, labels) in enumerate(train_loader):
#         imgs, labels = imgs.to(DEVICE), labels.to(DEVICE)
#         out = model(imgs)
#         loss = criterion(out, labels) / ACCUMULATION_STEPS
#         loss.backward()
#         if (i + 1) % ACCUMULATION_STEPS == 0:
#             optimizer.step()
#             scheduler.step()
#             optimizer.zero_grad()
#         correct += (out.argmax(1) == labels).sum().item()
#         total += labels.size(0)
#         if i % 50 == 0:
#             print(f"  Batch {i}/{len(train_loader)} | acc {correct/total:.3f}", end='\r')
#     return correct / total

# @torch.no_grad()
# def val_epoch():
#     model.eval(); total, correct = 0, 0
#     for imgs, labels in val_loader:
#         imgs, labels = imgs.to(DEVICE), labels.to(DEVICE)
#         correct += (model(imgs).argmax(1) == labels).sum().item()
#         total += labels.size(0)
#     return correct / total

# best_acc = 0
# for epoch in range(1, EPOCHS+1):
#     tr = train_epoch()
#     va = val_epoch()
#     if va > best_acc:
#         best_acc = va
#         torch.save(model.state_dict(), 'best_model.pth')
#     print(f"Epoch {epoch:02d} | train {tr:.4f} | val {va:.4f} | best {best_acc:.4f}")

In [ ]:
# import gc
# del model, optimizer, scheduler
# torch.cuda.empty_cache()
# gc.collect()

# Training Pipeline

In [ ]:
import torch
import random
import torchvision.transforms as transforms
from torchvision.datasets import ImageFolder
from torch.utils.data import DataLoader, random_split, Subset
import timm

In [ ]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(device)

cuda


In [ ]:
transform = transforms.Compose([
    transforms.Resize((128,128)),
    transforms.Lambda(lambda img: img.convert("RGB")),
    transforms.ToTensor(),
])

In [ ]:
dataset = ImageFolder('/content/khana', transform=transform)

In [ ]:
random.seed(42)

indices = list(range(len(dataset)))
random.shuffle(indices)

subset_size = int(0.3 * len(dataset))
subset_indices = indices[:subset_size]

dataset_small = Subset(dataset, subset_indices)

In [ ]:
train_size = int(0.8 * len(dataset))
val_size = len(dataset) - train_size

train_data, val_data = random_split(dataset, [train_size, val_size])

In [ ]:
train_loader = DataLoader(train_data, batch_size=64, shuffle=True, num_workers=4)
val_loader = DataLoader(val_data, batch_size=64, num_workers=2)

/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py:424: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  self.check_worker_number_rationality()


In [ ]:
model = timm.create_model('efficientnet_b0', pretrained=True, num_classes=80)
model = model.to(device)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:93: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


model.safetensors:   0%|          | 0.00/21.4M [00:00<?, ?B/s]

In [ ]:
from collections import Counter
import torch

labels = [label for _, label in dataset.samples]
class_counts = Counter(labels)

num_classes = len(dataset.classes)

weights = torch.zeros(num_classes)

for i in range(num_classes):
    weights[i] = 1.0 / class_counts[i]

weights = weights.to(device)

NameError: name 'dataset' is not defined

In [ ]:
criterion = torch.nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(model.parameters(), lr=5e-5, weight_decay=1e-4)

In [ ]:
print("Starting training...\n")

best_acc = 0

for epoch in range(10):   # fast run
    model.train()

    train_correct, train_total = 0, 0

    for images, labels in train_loader:
        images, labels = images.to(device), labels.to(device)

        outputs = model(images)
        loss = criterion(outputs, labels)

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        _, preds = torch.max(outputs, 1)
        train_correct += (preds == labels).sum().item()
        train_total += labels.size(0)

    train_acc = train_correct / train_total

    # validation
    model.eval()
    val_correct, val_total = 0, 0

    with torch.no_grad():
        for images, labels in val_loader:
            images, labels = images.to(device), labels.to(device)
            outputs = model(images)

            _, preds = torch.max(outputs, 1)
            val_correct += (preds == labels).sum().item()
            val_total += labels.size(0)

    val_acc = val_correct / val_total

    if val_acc > best_acc:
        best_acc = val_acc
        torch.save(model.state_dict(), "best_model.pth")

    print(f"Epoch {epoch+1} done | Train Acc: {train_acc:.4f} | Val Acc: {val_acc:.4f}")

print("\nyipppeeee Training complete!")
print(f"Best Val Accuracy: {best_acc:.4f}")

Starting training...



/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py:432: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  self.check_worker_number_rationality()
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings

Epoch 1 done | Train Acc: 0.6807 | Val Acc: 0.8182


/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in by

Epoch 2 done | Train Acc: 0.8557 | Val Acc: 0.8514


/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in by

Epoch 3 done | Train Acc: 0.9076 | Val Acc: 0.8590


/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in by

Epoch 4 done | Train Acc: 0.9424 | Val Acc: 0.8631


/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in by

Epoch 5 done | Train Acc: 0.9637 | Val Acc: 0.8641


/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in by

Epoch 6 done | Train Acc: 0.9737 | Val Acc: 0.8672


/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in by

Epoch 7 done | Train Acc: 0.9808 | Val Acc: 0.8658


/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in by

Epoch 8 done | Train Acc: 0.9832 | Val Acc: 0.8616


/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(


Starting training...

/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py:432: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  self.check_worker_number_rationality()
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
Epoch 1 done | Train Acc: 0.6851 | Val Acc: 0.8194
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
Epoch 2 done | Train Acc: 0.8555 | Val Acc: 0.8528
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
Epoch 3 done | Train Acc: 0.9079 | Val Acc: 0.8611
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
Epoch 4 done | Train Acc: 0.9434 | Val Acc: 0.8642
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
Epoch 5 done | Train Acc: 0.9637 | Val Acc: 0.8638
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
Epoch 6 done | Train Acc: 0.9757 | Val Acc: 0.8641
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/PIL/Image.py:1047: UserWarning: Palette images with Transparency expressed in bytes should be converted to RGBA images
  warnings.warn(


## Trying test data

In [ ]:
class_names = dataset.classes

In [ ]:
import os
import random
import shutil

test_folder = '/content/test_images'
os.makedirs(test_folder, exist_ok=True)

In [ ]:
data_path = '/content/khana'

all_images = []

for cls in os.listdir(data_path):
    cls_path = os.path.join(data_path, cls)

    if not os.path.isdir(cls_path):
        continue

    for img in os.listdir(cls_path):
        if img.startswith('.'):   # 🔥 skip junk
            continue
        all_images.append(os.path.join(cls_path, img))

# shuffle and pick 30
random.shuffle(all_images)
test_samples = all_images[:30]

In [ ]:
for img_name in os.listdir(test_folder):

    # skip junk files
    if img_name.startswith('.'):
        continue

    img_path = os.path.join(test_folder, img_name)

    try:
        img = Image.open(img_path)
        pred = predict_image(img_path)

        print(f"{img_name} → {pred}")

    except:
        print(f"Skipping bad file: {img_name}")

In [ ]:
import os

test_folder = '/content/test_images'

for file in os.listdir(test_folder):
    if file.startswith('.') or file.startswith('._'):
        os.remove(os.path.join(test_folder, file))

In [ ]:
for img_path in test_samples:
    shutil.copy(img_path, test_folder)

In [ ]:
from PIL import Image

def predict_image(img_path):
    img = Image.open(img_path)
    img = transform(img).unsqueeze(0).to(device)

    with torch.no_grad():
        outputs = model(img)
        _, pred = torch.max(outputs, 1)

    return class_names[pred.item()]

In [ ]:
import matplotlib.pyplot as plt

for img_name in os.listdir(test_folder):
    img_path = os.path.join(test_folder, img_name)

    img = Image.open(img_path)
    pred = predict_image(img_path)

    plt.imshow(img)
    plt.title(f"Prediction: {pred}")
    plt.axis('off')
    plt.show()

In [ ]:
correct = 0
total = 0

for img_path in test_samples:
    true_label = img_path.split('/')[-2]  # folder name
    pred = predict_image(img_path)

    if pred == true_label:
        correct += 1
    total += 1

print("Test Accuracy:", correct / total)

In [ ]:
from PIL import Image
img_path = '/content/images.jpeg'


def predict_image(img_path):
    img = Image.open(img_path)
    img = transform(img).unsqueeze(0).to(device)

    with torch.no_grad():
        outputs = model(img)
        _, pred = torch.max(outputs, 1)

    return class_names[pred.item()]

In [ ]:
import matplotlib.pyplot as plt

img = Image.open(img_path)
pred = predict_image(img_path)

plt.imshow(img)
plt.title(f"Prediction: {pred}")
plt.axis('off')
plt.show()

In [ ]:
import os
import random
import shutil

test_folder = '/content/test_food'
os.makedirs(test_folder, exist_ok=True)

In [ ]:
import os

test_folder = '/content/test_food'

for file in os.listdir(test_folder):
    if file.startswith('.') or file.startswith('._'):
        os.remove(os.path.join(test_folder, file))

In [ ]:
import os

for file in os.listdir(test_folder):
    if file.startswith('.') or file.startswith('._'):
        os.remove(os.path.join(test_folder, file))

In [ ]:
from PIL import Image

def predict_image(img_path):
    img = Image.open(img_path)
    img = transform(img).unsqueeze(0).to(device)

    with torch.no_grad():
        outputs = model(img)
        probs = torch.nn.functional.softmax(outputs, dim=1)
        conf, pred = torch.max(probs, 1)

    return class_names[pred.item()], conf.item()

In [ ]:
import matplotlib.pyplot as plt

for img_name in os.listdir(test_folder):
    img_path = os.path.join(test_folder, img_name)

    try:
        img = Image.open(img_path)
        pred, conf = predict_image(img_path)

        plt.imshow(img)
        plt.title(f"{pred} ({conf*100:.1f}%)")
        plt.axis('off')
        plt.show()

    except:
        print(f"Skipping: {img_name}")

In [ ]:
labels = {
    "img1.jpg": "idli",
    "img2.jpg": "aloo methi",
    "img3.jpg": "sev puri",
    "img4.jpg": "samosa",
    "img5.jpg": "aloo paratha",
    "img6.jpg": "paneer masala",
    "img7.jpg": "chana masala",
    "img8.jpg": "paneer masala",
    "img9.jpg": "gulab jamun",
    "img10.jpg": "kulfi",
    "img11.jpg": "chaas",
    "img12.jpg": "fish curry",
    "img13.jpg": "rasgulla",
    "img14.jpg": "samosa",
    "img15.jpg": "chicken wings",
    "img16.jpg": "vada pav"
}

In [ ]:
correct = 0
total = 0

for img_name in os.listdir(test_folder):
    if img_name.startswith('.'):
        continue

    if img_name not in labels:
        continue

    img_path = os.path.join(test_folder, img_name)

    pred, _ = predict_image(img_path)
    true = labels[img_name]

    if pred == true:
        correct += 1

    total += 1

print(f"Accuracy: {correct}/{total} = {correct/total:.4f}")

# QUESTION 2

In [ ]:
!pip install ultralytics

In [ ]:
import torchvision.transforms as transforms

In [ ]:
from ultralytics import YOLO

detector = YOLO("yolov8n.pt")

In [ ]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

model = timm.create_model('efficientnet_b0', pretrained=False, num_classes=80)
model.load_state_dict(torch.load("best_model.pth"))
model = model.to(device)
model.eval()

In [ ]:
from torchvision.datasets import ImageFolder

dataset = ImageFolder('/content/khana', transform=transform)
class_names = dataset.classes

In [ ]:
def classify_crop(crop):
    crop = cv2.cvtColor(crop, cv2.COLOR_BGR2RGB)
    crop = Image.fromarray(crop)
    crop = transform(crop).unsqueeze(0).to(device)

    with torch.no_grad():
        outputs = model(crop)
        _, pred = torch.max(outputs, 1)

    return class_names[pred.item()]

In [ ]:
thali_path = '/content/khana/thali'

In [ ]:
thali_path = '/content/khana/thali'

for img_name in os.listdir(thali_path):

    if img_name.startswith('.'):
        continue

    img_path = os.path.join(thali_path, img_name)

    print(f"Processing: {img_name}")

    image = cv2.imread(img_path)
    image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)

    results = detector(img_path)
    boxes = results[0].boxes.xyxy.cpu().numpy()

    for box in boxes:
        x1, y1, x2, y2 = map(int, box)

        crop = image[y1:y2, x1:x2]

        if crop.size == 0:
            continue

        try:
            label = classify_crop(crop)
        except:
            continue

        cv2.rectangle(image_rgb, (x1,y1), (x2,y2), (0,255,0), 2)
        cv2.putText(image_rgb, label, (x1, y1-10),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0,255,0), 2)

    # 🔥 SAVE IMAGE
    save_path = os.path.join(output_folder, img_name)
    cv2.imwrite(save_path, cv2.cvtColor(image_rgb, cv2.COLOR_RGB2BGR))